### PR TITLE
Supported Nested Modification of Data Structures

### PR DESCRIPTION
**Pull request by b-butler**
## Description

Adds support for a read, modify, write model of modification automatically with HOOMD's attributes even with syncing to C++. This involves various other changes. These are relevant in that they arise from bug fixes with increased testing and changes to interal APIs.

## Motivation and context

This is to prevent users from modifying copies where changes do not persist.

## How has this been tested?

Many tests have been added for the changes.

## Change log

```
Added: HOOMD attributes support nested modification.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

**Review by b-butler — COMMENTED** [review]
- [x] Add documentation to `ARCHITECTURE.md`
- [x] Add isolation tests to `TypeParameterDict`
- [ ] Add isolation tests to `ParameterDict`

File: hoomd/update/box_resize.py [hunk 1]
```diff
@@ -55,17 +63,17 @@ class BoxResize(Updater):
...
+                               box2=Box,
                                variant=Variant,
                                filter=ParticleFilter)
         params['box1'] = box1
         params['box2'] = box2
         params['variant'] = variant
         params['trigger'] = trigger
         params['filter'] = filter
+        for attr in ("box1", "box2"):
+            params._set_special_getset(attr, _box_getter, _box_setter)
```
Comment (joaander) [inline hoomd/update/box_resize.py:76]: Does this fix #705?

File: hoomd/pytest/test_simulation.py [hunk 2]
```diff
@@ -333,8 +333,9 @@ def test_run_limit(simulation_factory, lattice_snapshot_factory):
         sim.run(-1)
 
 
-def test_seed(simulation_factory, lattice_snapshot_factory):
-    sim = simulation_factory()
```
Comment (joaander) [inline hoomd/pytest/test_simulation.py:337]: Why this change? `simulation_factory` takes care of some important things like reducing the sorter grid size to run unit tests faster. While the sorter is never activated in this test, this pattern may get copied and pasted elsewhere.

File: hoomd/conftest.py [hunk 3]
```diff
@@ -71,6 +71,7 @@ def make_simulation(snapshot=None, domain_decomposition=None):
                 sim.create_state_from_snapshot(snapshot)
             else:
                 sim.create_state_from_snapshot(snapshot, domain_decomposition)
+        sim.seed = numpy.random.randint(100_000)
```
Comment (joaander) [inline hoomd/conftest.py:74]: I see, this is the change is why the see test no longer users `simulation_factory`. Use a fixed seed so that failures are as repeatable as possible.

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks! This is a very complete effort to provide an abstract API that operates the way users expect. There are a few changes I would like to see to the test rig, but otherwise this is good to merge.

File: hoomd/update/box_resize.py (hunk 1 above)
Comment (b-butler) [reply hoomd/update/box_resize.py:76]: No idea but I could look into it.

File: hoomd/pytest/test_simulation.py (hunk 2 above)
Comment (b-butler) [reply hoomd/pytest/test_simulation.py:337]: Yes this is to prevent developers from not setting a seed when they should and avoid testing error messages.

File: hoomd/update/box_resize.py (hunk 1 above)
Comment (joaander) [reply hoomd/update/box_resize.py:76]: No problem, we can solve in another PR if appropriate. Though it is related in spirit to the changes you mare making in this PR.

Comment (b-butler) [reply hoomd/update/box_resize.py:76]: I have verified that currently this is still a bug.

**Review by joaander — APPROVED** [review]
Thanks!
